records/2024-08-14.suspend-resume-callbacks.md

```diff
@@ -0,0 +1,117 @@
+A -. No .-> B1
+
+B -- Yes --> B1((END))
+B == No ==> D{Callbacks reaching\nmax retries?}
+
+D == Yes ==> DYes1[Trigger critical alarm] ==> DYes2[Manually suspend callbacks] ==> DYes3[Email service about suspension] ==> E{Has service owner\ncontacted us?}
+D -. No .-> F{24h elapsed?}
+
+E == Yes ==> EY1[Manually resume callbacks] --> EY2((END))
+E -. Wait .-> E
+
+F -. No .-> B
+F -. Yes .-> A
+
+style START fill::#bbf,stroke:#fff,stroke-width:2px,color:#fff,stroke-dasharray: 5 5
+style B1 fill::#,stroke:#E35,stroke-width:2px,color:#fff,stroke-dasharray: 5 5
+style EY2 fill::#bbf,stroke:#E35,stroke-width:2px,color:#fff,stroke-dasharray: 5 5
+
+```
+
+### Implementation overview
+To keep track of a service's suspension status and when the suspension took place we will add two new columns to the `service_callback_api` table:
+
+- `suspended`
+- `suspended_at`
+
+When the platform admin UI is used to suspend a service's callback, we will set `suspended` to `True` and record the time in the `suspended_at` column. By storing these values we facilitate tracking of suspended services so they can be easily displayed in the platform admin UI to be quickly resumed after the service resolves the issues with their API and contacts us.
+
+Should we automate the suspension and resumption process later down the road, the `suspended_at` column will provided us with the context we need to determine when we can automatically resume a service's callbacks. (e.g Time has reached a pre-defined threshold and we resume their callbacks.)
+
+## Additional considerations
+
+### The burden of manual intervention
+As of now the frequency at which callback API's go down during larger sends, leading strain on the system, is unknown. Should this be a rare occurrence, then manual intervention should suffice as a course of action.
+
+However if this issue occurs more frequently, particularly as Notify grows, then we should consider automating the suspension and resumption process in order to eliminate the burden of manual intervention, and maintain our velocity as a team.
+
+## Future considerations
+### Maintaining data quality as Notify grows
+Completeness, consistency, reliability and timeliness are pillars of data quality. While service owners are responsible for maintaining uptime of their API, in the case of their service going down, it's fair to assume the onus is on them to improve their uptime.
+
+As of today, there are only 39 services making use of callbacks. As Notify grows it may be of greater importance to understand how services are using callback data and how integral it is to their operational requirements.
+
+When we intervene and choose to stop the delivery of callbacks to a service, we potentially trade a service's data quality for system performance. After adequate research on how services use this data, we should consider and relate these findings to any potential reputational and policy implications of choosing to suspend a service.
+
```

Comment:
```suggestion

### Detaching the callback handling task from thread-bound http calls constraint

At the time of this writing, our Celery task that handles API callbacks are bound to one process/thread as these run on a Python implementation with the Global Interpreter Lock. To alleviate the risk of a scenario where a failing callbacks constantly times out and locks all process/threads available to the Celery callback handling tasks, we could move away from a synchronous/thread-bound model.

1. Use a Python asynchronous solution that unlocks the thread to process other network calls until the HTTP calls timeout. For this solution, this would need to play well with the Celery framework to preserve its retry feature.
1. Use a Python asynchronous solution that unlocks the thread to process other network calls until the HTTP calls timeout. For this solution, we would move away from Celery but keep on using Python. We need to preserve the retry features that Celery natively offers without using Celery.
1. We can move that piece of logic into AWS lambda. Starting a lambda would be similar to a process/thread except that we can launch as many as we want and these would handle high spikes pretty well typical to notification bursts that GCNotify experiences. We need to preserve the retry features that Celery natively offers without using Celery.
1. Stay within the Kubernetes space but move to a technology different than Python such as NodeJS or other. NodeJS would be a good candidate as it naturally queues up waiting green threads in the background until it expires/times out. We need to preserve the retry features that Celery natively offers without using Celery.

Tackling this issue would allow us to tackle callbacks faster and more reliably, without the impact to other services and other GCNotify's components. The users would still need to manage their callback APIs properly but that would highly reduce the risk when they are not doing it.
```